sdk/spring/azure-spring-integration-servicebus/src/main/java/com/azure/spring/integration/servicebus/converter/ServiceBusMessageConverter.java

```diff
@@ -77,59 +83,95 @@ protected IMessage fromByte(byte[] payload) {
     @Override
-    protected void setCustomHeaders(MessageHeaders headers, IMessage serviceBusMessage) {
-
-        if (headers.containsKey(MessageHeaders.CONTENT_TYPE)) {
-            Object contentType = headers.get(MessageHeaders.CONTENT_TYPE);
-
-            if (contentType instanceof MimeType) {
-                serviceBusMessage.setContentType(((MimeType) contentType).toString());
-            } else {
-                serviceBusMessage.setContentType((String) contentType);
-            }
-        }
-
-        if (headers.containsKey(MessageHeaders.ID)) {
-            serviceBusMessage.setMessageId(String.valueOf(headers.get(MessageHeaders.ID, UUID.class)));
-        }
-
-        if (headers.containsKey(MessageHeaders.REPLY_CHANNEL)) {
-            serviceBusMessage.setReplyTo(headers.get(MessageHeaders.REPLY_CHANNEL, String.class));
-        }
-
-        if (headers.containsKey(AzureHeaders.SCHEDULED_ENQUEUE_MESSAGE)) {
-            Integer integerValue = headers.get(AzureHeaders.SCHEDULED_ENQUEUE_MESSAGE, Integer.class);
-            if (null != integerValue) {
-                serviceBusMessage.setScheduledEnqueueTimeUtc(Instant.now().plus(Duration.ofMillis(integerValue)));
-            }
-        }
+    protected void setCustomHeaders(MessageHeaders headers, IMessage message) {
+
+        // Spring MessageHeaders
+        getStringHeader(headers, MessageHeaders.ID).ifPresent(message::setMessageId);
+        getStringHeader(headers, MessageHeaders.CONTENT_TYPE).ifPresent(message::setContentType);
+        getStringHeader(headers, MessageHeaders.REPLY_CHANNEL).ifPresent(message::setReplyTo);
+
+        // AzureHeaders.
+        getStringHeader(headers, AzureHeaders.RAW_ID).ifPresent(message::setMessageId);
+        Optional.ofNullable(headers.get(AzureHeaders.SCHEDULED_ENQUEUE_MESSAGE, Integer.class))
+                .map(Duration::ofMillis)
+                .map(Instant.now()::plus)
+                .ifPresent(message::setScheduledEnqueueTimeUtc);
+
+        // ServiceBusMessageHeaders, service bus headers have highest priority.
+        getStringHeader(headers, MESSAGE_ID).ifPresent(message::setMessageId);
+        Optional.ofNullable((Duration) headers.get(TIME_TO_LIVE))
+                .ifPresent(message::setTimeToLive);
+        Optional.ofNullable((Instant) headers.get(SCHEDULED_ENQUEUE_TIME_UTC))
+                .ifPresent(message::setScheduledEnqueueTimeUtc);
+        getStringHeader(headers, SESSION_ID).ifPresent(message::setSessionId);
+        getStringHeader(headers, CORRELATION_ID).ifPresent(message::setCorrelationId);
+        getStringHeader(headers, TO).ifPresent(message::setTo);
+        getStringHeader(headers, LABEL).ifPresent(message::setLabel);
+        getStringHeader(headers, REPLY_TO_SESSION_ID).ifPresent(message::setReplyToSessionId);
+        getStringHeader(headers, PARTITION_KEY).ifPresent(message::setPartitionKey);
+        getStringHeader(headers, VIA_PARTITION_KEY).ifPresent(message::setViaPartitionKey);
```

Comment:
What's this VIA_PARTITION_KEY used for?

I think we should also keep an eye on the track2 SB sdk, in case we don't add a header which is no longer used in the new sdk. https://github.com/Azure/azure-sdk-for-java/blob/master/sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusMessage.java.

Reply:
VIA_PARTITION_KEY and LABEL are not used in the track2 sdk.

Reply:
Good point. Updated.